test/simulator/abilities/dancer.js

```diff
@@ -52,4 +53,60 @@ describe('Dancer', function () {
 		assert.fainted(battle.p2.active[0]);
 		assert.fainted(battle.p2.active[1]);
 	});
+
+	it('should not copy a move that failed or was blocked by Protect', function () {
+		battle = common.createBattle({gameType: 'doubles'}, null, new PRNG([1, 2, 3, 4]));
+		const p1 = battle.join('p1', 'Guest 1', 1, [
+			{species: 'Oricorio', level: 98, ability: 'dancer', item: 'laggingtail', moves: ['dragondance', 'protect', 'teeterdance']},
+			{species: 'Oricorio', level: 99, ability: 'dancer', moves: ['featherdance']},
+		]);
+		const p2 = battle.join('p2', 'Guest 2', 1, [
+			{species: 'Oricorio', ability: 'dancer', moves: ['fierydance', 'protect', 'teeterdance']},
+			{species: 'Shedinja', ability: 'wonderguard', moves: ['finalgambit']},
+		]);
+		p1.active[0].boostBy({atk: 6, spe: 6});
+		p2.active[0].boostBy({atk: -6});
+		p2.active[1].boostBy({spe: 6});
+		p1.chooseMove(1).chooseMove(1, 1);
+		p2.chooseMove(1, -2).chooseMove(1, 1);
+		assert.fullHP(p2.active[0]);
+		assert.statStage(p1.active[0], 'atk', 6);
+		assert.statStage(p1.active[1], 'atk', 0);
+		assert.statStage(p2.active[0], 'atk', -6);
+		assert.statStage(p2.active[0], 'spe', 0);
+		// Next turn
+		battle.choose('p1', 'move 1');
+		battle.choose('p2', 'move 2');
+		battle.commitDecisions();
+		assert.statStage(p1.active[0], 'atk', 6);
+		assert.statStage(p1.active[1], 'atk', 0);
+		// Next turn: Teeter Dance should be copied as long as it hits one thing
+		battle.choose('p1', 'move 2');
+		battle.choose('p2', 'move 3');
+		battle.commitDecisions();
+		assert.fullHP(p1.active[1]);
+		// Next turn: Teeter Dance should NOT be copied if everything it hits is already confused
+		battle.choose('p1', 'move 3');
+		assert.constant(() => p1.active[0].volatiles['confusion'], () => battle.commitDecisions());
+	});
+
+	it('should not copy a move that missed', function () {
+		battle = common.createBattle();
+		const p1 = battle.join('p1', 'Guest 1', 1, [{species: 'Oricorio', ability: 'dancer', item: 'choicescarf', moves: ['revelationdance']}]);
+		const p2 = battle.join('p2', 'Guest 2', 1, [{species: 'Oricorio', ability: 'dancer', item: 'brightpowder', moves: ['dig']}]);
+		p1.active[0].boostBy({accuracy: -6});
+		p2.active[0].boostBy({evasion: 6});
+		battle.commitDecisions();
```

Comment:
Is this guaranteed to miss, or why aren't you using a specific seed?

Reply:
There's an assert later that checks to be sure it did miss, but I can make this use a set seed if PRNG might change how it produces seeds.

Reply:
...That's not how asserts work.

Reply:
Asserts say "this is true, or there's a bug in the code". Not "this is true, or we got unlucky by the RNG".